lib/cdo/newrelic.rb

```diff
@@ -0,0 +1,141 @@
+    assign_alert_policy(DISABLED_ALERT_POLICY_ID, server_names)
+  end
+
+  # Disables alerts to the given server names.
+  def enable_alerts(server_names)
+    assign_alert_policy(PRODUCTION_ALERT_POLICY_ID, server_names)
+  end
+
+  # Disables alerts for the given servers
+  # param [Array<String>] server_names A list of new relic server names (not ids).
+  def assign_alert_policy(policy_id, server_names)
+    policy = alert_policy(policy_id)
+    map = server_name_to_id_map
+
+    server_names.each do |name|
+      server_id = map[name]
+      if server_id
+        policy['links']['servers'] << server_id
+      else
+        next
+      end
+    end
+
+    body = {'alert_policy': policy}.to_json
+    call_newrelic_rest("alert_policies/#{policy_id}.json", 'PUT', body)
+  end
+
+  # Returning a map from alert policy id to alert hash as described at
+  # https://rpm.newrelic.com/api/explore/alert_policies/list
+  def get_alert_policies_map
+    policies = call_newrelic_rest('alert_policies.json')['alert_policies']
+    index(policies, 'id')
+  end
+
+  # Returning a hash describing the specified alert policies as described at
+  # https://rpm.newrelic.com/api/explore/alert_policies/show
+  def alert_policy(alert_policy_id)
+    call_newrelic_rest("alert_policies/#{alert_policy_id}.json")['alert_policy']
+  end
+
+  # Returns a hash describing the disabled server alert policy.
+  def disabled_alert_policy
+    alert_policy(DISABLED_ALERT_POLICY_ID)
+  end
+
+  # Returns a hash describing the production server alert policy.
+  def production_alert_policy
+    alert_policy(PRODUCTION_ALERT_POLICY_ID)
+  end
+
+  # Invokes a NewRelic REST action, parsing the JSON result and returning a hash.
+  # @throws a runtime exception if an HTTP or JSON parsing error occurs
+  def call_newrelic_rest(path, method = 'GET', body = nil)
+    uri = newrelic_uri(path)
+
+    case method.upcase
+    when 'GET'
+      req = Net::HTTP::Get.new(uri)
+    when 'PUT'
+      req = Net::HTTP::Put.new(uri)
+    when 'POST'
+      req = Net::HTTP::Post.new(uri)
+    else
+      raise "Unknown method #{method}"
+    end
+
+    req['X-Api-Key'] = api_key
+    req['Content-Type'] = 'application/json'
+
+    req.body = body if body
+
+    http = Net::HTTP.new(uri.hostname, uri.port)
+    http.use_ssl = true
+
+    response = http.start { http.request(req) }
+
+    if response.code == "200"
+      JSON::parse(response.body)
+    else
+      raise "HTTP Error #{response.code}: #{response.message} #{response.body}"
+    end
+  end
+
+  def newrelic_uri(path)
+    URI("#{NEWRELIC_URL}/#{path}")
+  end
+
+  # "Index" an array of hashes by return a map from the value of the given field
+  # to the corresponding record. Field should be unique to avoid collisions.
+  def index(array, field)
+    {}.tap do |result|
+      array.each do |item|
+        raise "#{field} is not unique" unless result[item[field]].nil?
+        result[item[field]] = item
+      end
+    end
+  end
```

Comment:
`#group_by` could be used here:
```
array = [{id: 1, x: 'a'}, {id: 2, x: 'b'}]
> array.group_by{|x|x[:id]}
=> {1=>[{:id=>1, :x=>"a"}], 2=>[{:id=>2, :x=>"b"}]}
```
but to enforce uniqueness and collapse the value-array into a single object you'd need to do some extra steps:
```
raise "#{field} is not unique" unless array.uniq!{|x|x[field]}.nil?
array.group_by{|x|x[field]}.map{|k,v| [k,v.first]}.to_h
```
or using [`Hash#transform_values`](https://github.com/rails/rails/blob/4-2-stable/activesupport/lib/active_support/core_ext/hash/transform_values.rb) extension from ActiveSupport:
```
require 'active_support/core_ext/hash/transform_values'
raise "#{field} is not unique" unless array.uniq!{|x|x[field]}.nil?
array.group_by{|x|x[:id]}.transform_values{|v|v.first}
```
so not much more concise in the end